verify_rpms/rpm_verifier.py

```diff
@@ -106,42 +108,107 @@ def get_signed_rpms_keys(rpms: list[str]) -> list[str]:
-        )
-    return failure, output.lstrip()
+        results["keys"] = dict(Counter(signed_rpms_keys).most_common())
+        results["keys"]["unsigned"] = len(unsigned_rpms)
+    return results
 
 
-def generate_results(
-    processed_image: ProcessedImage,
+def inspect_image_ref(
+    image_url: str, image_digest: str, runner: Callable = run
 ) -> dict[str, Any]:
     """
-    Generate the results that should be added to the results of the task
-    :param processed_image: ProcessedImage that contains all the rpms information
-    :returns: dictionary with results for the image
+    Inspect the image reference by running Skopeo
+    :param image_url: image url to inspect
+    :param image_digest: image digest to inspect
+    :param runner: subprocess.run to run CLI commands
+    :return: dictionary containing the inspection details
     """
-    results: dict[str, Any] = {}
-    if processed_image.error != "":
-        results["error"] = processed_image.error
+    image_with_digest = f"docker://{image_url.split(':')[0]}@{image_digest}"
+    inspect = runner(
+        [  # pylint: disable=duplicate-code
+            "skopeo",
+            "inspect",
+            "--raw",
+            image_with_digest,
+        ],
+        capture_output=True,
+        text=True,
+        check=True,
+    )
+    return json.loads(inspect.stdout)
+
+
+def generate_image_output(image: str, unsigned_rpms: list[str], error: str) -> str:
+    """
+    Generates output for each container according to the scan
+    :param image: image name
+    :param unsigned_rpms: list of unsigned RPMs
+    :param error: error string
+    :return: output string
+    """
+    header = f"Image: {image}\n"
+    if error != "":
+        output = f"{header}Error occurred:\n{error}\n"
+    elif unsigned_rpms:
+        output = f"{header}Found unsigned RPMs:\n{unsigned_rpms}\n"
     else:
-        results["keys"] = dict(Counter(processed_image.signed_rpms_keys).most_common())
-        results["keys"]["unsigned"] = len(processed_image.unsigned_rpms)
-    return results
+        output = f"{header}No unsigned RPMs found\n"
+    return output
+
+
+def get_images_from_inspection(
+    inspect_results: dict[str, Any], image_url: str, image_digest: str
+) -> list[str]:
+    """
+    Analyze the inspection results and return all the images for the image reference
+    The image reference may refer to an image index (AKA Manifest list)
+    In this case, extract all the digests from the manifests and return the list of image references
+    :param inspect_results: inspection results dictionary
+    :param image_url: image url to inspect
+    :param image_digest: image digest to inspect
+    :return: list of images to scan
+    """
+    if "manifests" in inspect_results:
+        manifests = inspect_results["manifests"]
+        image_list = [
+            f"{image_url.split(':')[0]}@" + man["digest"] for man in manifests
```

Comment:
use the -1 index instead of 0 to support image references with a port, for example
registry.local:5000/myimage:latest

this should be fixed in multiple places in this script.

Reply:
@gbenhaim Do all image-urls comes with port number?

Reply:
Done